exercises/firmware_auditing/README.md

```diff
@@ -0,0 +1,318 @@
+By using the built-in `sum` function, `allocator_capabilities`, and `unique_allocator_compartments`, we can define an object that maps from a given compartment to its total allocation quota.
+
+Following from this, we can define our second allocation limiting rule: `all_compartments_allocate_leq(limit)`.
+This function checks that no individual compartment can allocate memory greater than a given limit at one time, across all of its allocator capabilities.
+This hence ensures that only a certain amount of memory can be used by one compartment at any given time.
+
+For our final allocation limiting rule, we first define a helper rule `total_quota` which sums up the quota of every allocator capability in the firmware.
+We use this to construct the final check `total_allocation_leq(limit)`, which ensures that only a certain amount of memory can be used by firmware at any one time.
+This can be useful information for auditing firmware running on systems with limited memory resources, or with multiple processes running simultaneously.
+
+As in our other examples, we finish by making a `valid` rule which evaluates to a Boolean, which audits whether our firmware image is valid.
+Using our 3 functions, we can easily set custom allocation limits.
+For this exercise, we decide that all sealed allocator capabilities must be valid, all individual allocations must be at most 20 KiB, no compartment must allocate more than 40 KiB at once, and the entire firmware must not allocate more than 100 KiB at once.
+
+We can audit our firmware using the following command:
+```sh
+cheriot-audit --board=cheriot-rtos/sdk/boards/sonata.json \
+              --firmware-report=build/cheriot/cheriot/release/firmware_auditing_part_3.json \
+              --module=exercises/firmware_auditing/part_3/malloc_check.rego \
+              --query='data.malloc_check.valid'
+```
+
+In this instance, the output of this test should be `true`, as our defined firmware meets these properties.
+You can check this yourself by looking at the source files and the values of the intermediary rules.
+To test that the policy is working, you can either change the amount of memory allocated by the firmware (making sure to rebuild), or change the policy itself to enforce lower limits.
+For example, changing the line
+```
+all_allocations_leq(20480)  # 20 KiB
+```
+to the new line
+```
+all_allocations_leq(10240)  # 10 KiB
+```
+should cause the `valid` rule of the policy to evaluate to `false`, because the `malloc_many` compartment contains a capability that permits the allocation of 16 KiB at once, which is greater than our specified limits.
+
+Try playing around with these values to convince yourself that the policy is working as we expect it to.
+
+## Beyond these exercises
+
+There are other pieces of information that the linker exposes, which we do not use in this exercise. For example:
+ - You could check that only certain compartments access certain MMIO capabilities.
```

Comment:
Maybe link to part 3 of the other exercise here. *You could even not mention it at the beginning and only here.*